File: Python/Basic Recursion/factorial.py
def method_1(n):
    '''
        Recursive approach
        Time Complexity = O(n)
        Space Complexity = O(n)
    '''
    if n == 1:
        return 1

    return n * method_1(n - 1)


def method_2(n):
    '''
        Iterative approach
        Time Complexity = O(n)
        Space Complexity = O(1)
    '''
    factorial = 1
    for i in range(1, n + 1):
        factorial *= i

    return factorial


print("Enter the number: ")
n = int(input())
factorial = method_1(n)
print("Method 1 Result:", factorial)
factorial = method_2(n)
print("Method 2 Result:", factorial)

File: C++/Basic Recursion/factorial.cpp
#include <bits/stdc++.h>

using namespace std;

long long method_1(long long n)
{
    /*
        Recursive approach
        Time Complexity = O(n)
        Space Complexity = O(n)
    */
    if (n == 1)
        return 1;

    return n * method_1(n - 1);
}

long long method_2(long long n)
{
    /*
        Iterative approach
        Time Complexity = O(n)
        Space Complexity = O(1)
    */
    long long factorial = 1;
    for (int i = 1; i <= n; i++)
    {
        factorial *= i;
    }

    return factorial;
}

int main()
{
    long long n;
    long long factorial;
    cout << "Enter the number: " << endl;
    cin >> n;
    factorial = method_1(n);
    cout << "Method 1 Result: " << factorial << endl;
    factorial = method_2(n);
    cout << "Method 2 Result: " << factorial << endl;

    return 0;
}